### PR TITLE
Pass "signed" data with the action to the agent

### PR DESCRIPTION
**Pull request by aleksmaus**
## What is the problem this PR solves?

Extends the Fleet Server actions schema in order to support signed actions passing to the agent as a part of the agent tamper protection.

## How does this PR solve the problem?

Passing the "signed" property with actions upon check-in.

Tested for backwards compatibility, confirmed that the older agent that are not aware of this feature ignore the new "signed" property, and successfully process actions.

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

The related issue is currently in the private repository.
- Requires Elasticsearch .fleet-actions mapping update PR https://github.com/elastic/elasticsearch/pull/93802. 

## Screenshots&Logs&Etc

Confirmed fleet server passes the "signed" property in raw actions response in the agent:
Screenshot 2023-02-15 at 3 56 50 PM

```
{
    "ack_token": "SWvdVoYBrGHKScX-NT6O",
    "action": "checkin",
    "actions": [
        {
            "agent_id": "24023d72-d4d5-4a19-95e9-3ebe699395e8",
            "created_at": "2023-02-15T20:35:08.244Z",
            "expiration": "2023-02-15T20:58:08.244Z",
            "data": {
                "id": "6a1d4a16-a9f5-4edb-96eb-ddaf9b3786c8",
                "query": "select * from osquery_info"
            },
            "id": "d1e02ee0-fc14-4930-8e55-f962e76a4895",
            "type": "INPUT_ACTION",
            "input_type": "osquery",
            "signed": {
                "data": "eyJAdGltZXN0YW1wIjoiMjAyMy0wMS0xMlQxNzowNzo0My4yNzhaIiwiYWN0aW9uX2lkIjoiZGIyNjNmNTQtZGQ3Yy00YWM4LTgxZmQtN2JkZjdjY2M5OWRmIiwiYWdlbnRzIjpbIjJhYmNlMjNiLTdiZDktNDc3Zi1iNmY3LTJmZWExOWZmYTRkZiIsIjNhYmNlMjNiLTdiZDktNDc3Zi1iNmY3LWFiZWExOWZmYTRkZiIsIjRhYmNlMjNiLTdiZDktNDc3Zi1iNmY3LWRmZWExOWZmYTRkZiJdLCJkYXRhIjp7ImlkIjoiYTQxMzg4NDgtYjJjYi00ZWJlLWFlOWItYzViNjYzMmZiNDM4IiwicXVlcnkiOiJzZWxlY3QgKiBmcm9tIG9zcXVlcnlfaW5mbyJ9LCJleHBpcmF0aW9uIjoiMjAyMy0wMS0xMlQxNzoxMjo0My4yNzlaIiwiaW5wdXRfdHlwZSI6Im9zcXVlcnkiLCJ0eXBlIjoiSU5QVVRfQUNUSU9OIiwidXNlcl9pZCI6IjM4NDQxMTA1NDEifQ==",
                "signature": "U6NOg4ssxpFVe19OFWpuDozTds5P4HGo7YxxbJrLe6FDYe4xKEvVD0qRFdOeTJ50YvtT6NnNOh7xE2YBP8hcR9DzmB6xMUL0EaeHzJZ7BOBAXKvD2IgJerp4ZyNdcUwFFYAM2n5Ua1m/KzdI7Cu96O5e58m9U7bTteuaLKgYAtKZvcdBdT28OExS4h4jeE0g2Lam5u4XqI9Iw6fSoj/jLAVVtPIOtlbsP+zx5F7z7cEZ9SwOuwfkv5HrmkPCWpc8Wf1wQdm5N79xMbHdNb2qjOG297ov56odWv7iZ3u6Ph+9u+WBZI/OMcA1uttGHC07WMjJEsEA9i1tFhCEGoBHsg=="
            }
        }
    ]
}
```

Confirmed backwards compatibility, the action with the new "signed" property is still processed by the older agent, using osquery live query as example here:

Screenshot 2023-02-15 at 3 58 27 PM

### CODE REVIEW

**Review by pzl — COMMENTED** [review]
proably need to update `model/openapi.yml` with the field. Changelog entry coming?

Do we have any tests that try to deserialize an action string/[]byte with all or some of these black-box or optional fields populated, and verify the right contents end up in the struct to be passed along?

if nothing else, prevents some typos in the struct tags, by making sure it matches a json we expect. I don't think anything is validating the `json:"signed"` part

**Comment by aleksmaus** [comment]
> Do we have any tests that try to deserialize an action string/[]byte with all or some of these black-box or optional fields populated, and verify the right contents end up in the struct to be passed along?
> 
> if nothing else, prevents some typos in the struct tags, by making sure it matches a json we expect. I don't think anything is validating the `json:"signed"` part

I was thinking, same as with some other json.RawMessage fields in the model, the fleet server would be just passing the payload through to the agent, not spending any cpu cycles on parsing and validation. Similar what we would do with the policy "data.signed" for example. This is a contract between kibana and the agent. The parsing, validation and signature check will be performed by the agent, thus there will be the unit test coverage on the agent side when this feature is implemented.

**Comment by pzl** [comment]
>  the fleet server would be just passing the payload through to the agent

yes, agreed, it's not actually inspecting the contents or deserializing those parts of the object.

Question, though: if one were to edit the struct tag from `json:"signed"` to be (incorrectly): `json:"signature"` -- does that affect the payload going to agent? Does agent still get the original doc with the expected payload under `signed` property? Does fleet's parsing (by knowing the property name) matter for delivery? I don't actually know. But I'd be interested in the result.

If fleet needs to know the key (even if it leaves the contents as untouched `[]byte`s) to deliver these opaque payloads correctly, then we probably need at least some minimal tests around that

**Comment by aleksmaus** [comment]
Added ```model/openapi.yml```. Added actions serializations tests, covering new Signed attribute on the schema and on API model side.

**Review by pzl — APPROVED** [review]

**Review by michel-laterman — APPROVED** [review]
looks good I have mostly nitpicks.
It should be easy enough to update my codegen PR with these changes once it's merged

File: model/openapi.yml [hunk 1]
```diff
@@ -325,6 +325,29 @@ components:
           x-go-custom-tag: yaml:"timeout" # openapi-generator
           x-oapi-codegen-extra-tags: # oapi-codegen tags
             yaml: "timeout"
+        signed:
+          description: The optional action signing data.
+          type: object
+          format: int64
```
[the remark this answers is not in the document]
Comment (aleksmaus) [reply model/openapi.yml:331]: oops, copy-pasta, fixed

**Pull request merged**